### ISSUE TITLE
GEM stops reporting

### ISSUE THREAD

**Issue by ptr727** [jkeljo/greeneye-monitor#2]
Hi, I am not sure if this is a GEM issue, a greeneye-monitor issue, or a home assistant issue, but I'm starting here, and please point me elsewhere if needed.

Problem is that home assistant would stop reporting on GEM event data.

Config:
```
# https://www.home-assistant.io/integrations/logger
logger:
  default: warning
  logs:
    homeassistant.components.greeneye: debug
```

```
# https://www.home-assistant.io/integrations/greeneye_monitor
greeneye_monitor:
  port: 8000
  monitors:
    - serial_number: "01000181"
      channels:
        - number: 1
          name: gem_1_channel_1
        - number: 2
          name: gem_1_channel_2
    - serial_number: "01000172"
      pulse_counters:
        - number: 3
          name: gem_2_water
          counted_quantity: "gal"
          counted_quantity_per_pulse: 10
          time_unit: "min"
        - number: 4
          name: gem_2_gas
          counted_quantity: "cuft"
          counted_quantity_per_pulse: 10
          time_unit: "min"
```

Logs:
```
2019-10-07 07:02:30 WARNING (SyncWorker_37) [homeassistant.components.influxdb] Database host is not accessible due to 'HTTPConnectionPool(host='influxdb.home.insanegenius.net', port=8086): Max retries exceeded with url: /write?db=home_assistant (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x148cb249dd10>: Failed to establish a new connection: [Errno 113] Host is unreachable'))', please check your entries in the configuration file (host, port, etc.) and verify that the database exists and is READ/WRITE. Retrying again in 60 seconds.
2019-10-07 07:02:31 WARNING (MainThread) [homeassistant.setup] Setup of person is taking over 10 seconds.
2019-10-07 07:02:31 WARNING (MainThread) [homeassistant.setup] Setup of rachio is taking over 10 seconds.
2019-10-07 07:02:31 WARNING (MainThread) [homeassistant.setup] Setup of greeneye_monitor is taking over 10 seconds.
2019-10-07 07:02:31 WARNING (MainThread) [homeassistant.setup] Setup of influxdb is taking over 10 seconds.
2019-10-07 07:02:43 WARNING (MainThread) [homeassistant.setup] Setup of envisalink is taking over 10 seconds.
2019-10-07 07:02:43 WARNING (MainThread) [homeassistant.setup] Setup of ecobee is taking over 10 seconds.
2019-10-07 07:02:43 WARNING (MainThread) [homeassistant.setup] Setup of sabnzbd is taking over 10 seconds.
2019-10-07 07:02:44 ERROR (MainThread) [homeassistant.components.sabnzbd] Connection to SABnzbd API failed
2019-10-07 07:02:44 ERROR (MainThread) [homeassistant.components.deconz] Error connecting to deCONZ gateway at 192.168.1.28
2019-10-07 07:02:44 WARNING (MainThread) [homeassistant.config_entries] Config entry for deconz not ready yet. Retrying in 5 seconds.
2019-10-07 07:02:44 WARNING (MainThread) [homeassistant.components.weather] Setup of platform met is taking over 10 seconds.
2019-10-07 07:02:44 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 273, in _on_client_connected
    await conn
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 278, in _handle_packets
    await asyncio.coroutine(self._listener)(packet)
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 311, in _handle_packet
    for listener in self._listeners])
  File "/usr/local/lib/python3.7/asyncio/tasks.py", line 380, in wait
    raise ValueError('Set of coroutines/Futures is empty.')
ValueError: Set of coroutines/Futures is empty.
2019-10-07 07:02:44 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 273, in _on_client_connected
    await conn
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 278, in _handle_packets
    await asyncio.coroutine(self._listener)(packet)
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 311, in _handle_packet
    for listener in self._listeners])
  File "/usr/local/lib/python3.7/asyncio/tasks.py", line 380, in wait
    raise ValueError('Set of coroutines/Futures is empty.')
ValueError: Set of coroutines/Futures is empty.
2019-10-07 07:38:28 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 273, in _on_client_connected
    await conn
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 277, in _handle_packets
    async for packet in packets:
  File "/usr/local/lib/python3.7/site-packages/greeneye/streams.py", line 62, in __anext__
    codecs.encode(packet, 'hex')))
greeneye.packets.MalformedPacketException: Invalid header 0xfe in packet: b'fe'
2019-10-07 07:38:28 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 273, in _on_client_connected
    await conn
  File "/usr/local/lib/python3.7/site-packages/greeneye/monitor.py", line 277, in _handle_packets
    async for packet in packets:
  File "/usr/local/lib/python3.7/site-packages/greeneye/streams.py", line 62, in __anext__
    codecs.encode(packet, 'hex')))
greeneye.packets.MalformedPacketException: Invalid header 0x0 in packet: b'000000'
```

Note the invalid packet data reports occurred after the GEM already stopped reporting. It may still be an issue, but I suspect the task exceptions are the main cause.

I have 4 GEM's, I have only just started configuring two of them, long story, installed in 2012 when I built the house, CT leads are too short to reach cabinets holding GEM, building inspector did not want to see any CT leads, I ended up not using the GEM's until recently, when I got into HASS, and I want to use them with HASS.

GEM's are configured to use BIN48-NET-Time.
I only hooked up a few inputs, if I can get HASS working reliably with the GEM's, I'll wire the rest.

I am running home assistant in docker on unraid, using latest public release.
0.99.3, 20190919.1

Please let me know how I can help troubleshoot.

**Comment by jkeljo** [comment]
Wow 4 GEMs! That might be part of it, I think the Hass integration didn’t anticipate having more than one in a residential environment. If you configure only one of them to send packets to Hass, and put only that one in the Hass config, does that work?

Is yours reporting data for some period of time and then stopping, or just never getting any data at all?

I have seen mine periodically send bad packets, but Hass recovers once it sends good ones again.

Those first two exceptions I’ve seen before, though only on shutdown. I’ll get a fix out for those just to help narrow down, and look into how many single-device assumptions are in the Hass code.

**Comment by ptr727** [comment]
I have 4 but only two is currently configured, and both send data.
Data from both gets recorded, until the exception, then data stops getting recorded.
If I restart HASS, recording of data starts again.

**Comment by ptr727** [comment]
I noticed another issue today, my HASS was consuming 1 CPU core 100%, 12% CPU for the python process.
I restarted it a couple times, same issue.

I set debug logging to debug, and I got my log spammed with, by the time I managed to kill the container my log had reached near 800K lines:
```
2019-10-07 21:56:21 INFO (MainThread) [greeneye.streams] 192.168.1.243:2529: Connected
2019-10-07 21:56:21 DEBUG (MainThread) [greeneye.streams] Short packet
2019-10-07 21:56:21 DEBUG (MainThread) [greeneye.streams] Short packet
2019-10-07 21:56:21 DEBUG (MainThread) [greeneye.streams] Short packet
2019-10-07 21:56:23 DEBUG (MainThread) [greeneye.streams] Short packet
2019-10-07 21:56:23 DEBUG (MainThread) [greeneye.streams] Short packet
```

This symptoms may be similar to:
https://github.com/home-assistant/home-assistant/issues/25887

What was weird is that I normally have logging for greeneye at debug, but it was only once I enabled global debug logging that I noticed the short packet errors, with global at warning I do not get the short packet logs.
```
# https://www.home-assistant.io/integrations/logger
logger:
  default: debug
  logs:
    homeassistant.components.greeneye: debug
```

The short packet looks to come from PacketIterator(), and I assume it ends in an infinite loop waiting for the rest of the packet, that may never arrive, or the connection closed before the logic exits the loop.

I see the comment about TCP and PSH vs. UDP, and I've not taken a look at a packet dump, but I do know that I had to fiddle with the ethernet to serial bridge parameters to get the GEM to be responsive, see:
http://brultech.com/community/viewtopic.php?f=29&t=2015&p=13460#p13460

My GEM's are currently configured to send data every 10s, and ethernet bridge idle time is set to 3s. So I expect the GEM to connect every 10s as serial data arrives, send the data, wait 3s, disconnect, 7 seconds later connect, send, wait 3s, disconnect, repeat. 
If this continues to be an issue I will capture traffic.

**Comment by ptr727** [comment]
I debugged using dump_packet.py.

```
        while len(packet) != packet_format.size:
            packet.extend(
                await self._stream.read(packet_format.size - len(packet)))
            if len(packet) != packet_format.size:
                LOG.debug("Short packet")
```
len(packet) is 540, packet_format.size is 619, _stream,read() returns immediately without reading data, tight infinite loop.

I think the logic could include a closed connection signal, or testing the read result, or graceful waiting for data, or adding a timeout waiting for data.

I'm learning Python on the fly, and I'll capture TCP dumps and add them later.

**Comment by ptr727** [comment]
```
488	24.443442	192.168.1.243	192.168.1.113	TCP	60	2060 → 8000 [SYN] Seq=0 Win=2048 Len=0 MSS=1460
490	24.443746	192.168.1.243	192.168.1.113	TCP	60	2060 → 8000 [ACK] Seq=1 Ack=1 Win=2048 Len=0

493	24.474271	192.168.1.243	192.168.1.113	TCP	309	2060 → 8000 [PSH, ACK] Seq=1 Ack=1 Win=2048 Len=255
0000   fe ff 05 04 b9 02 00 00 00 00 01 00 00 00 00 01
0010   00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00
0020   00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00
0030   00 00 01 00 00 00 00 01 00 00 00 00 01 00 00 00
0040   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0050   00 00 00 00 00 8b 01 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 8f 02 00 00 00 00 00 00 00 00 6a 02
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 ae 01 00 00 00 00 00 00 00 00 e1 01 00 00
0090   00 33 01 00 00 00 00 00 00 00 00 00 00 00 00 00
00a0   4b 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00b0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00c0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00d0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00e0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00f0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

494	24.498506	192.168.1.243	192.168.1.113	TCP	309	2060 → 8000 [PSH, ACK] Seq=256 Ack=1 Win=2048 Len=255
0000   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0010   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0020   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0030   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0040   00 00 00 00 00 00 3d 00 00 00 00 00 00 00 00 00
0050   00 00 00 00 00 8f 02 00 00 00 00 00 00 00 00 6a
0060   02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 ae 01 00 00 00 00 00 00 00 00 e1 01 00
0080   00 00 33 01 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 4b 02 00 00 00 00 00 00 00 00 00 00 00 00 00
00a0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00b0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00c0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00d0   00 00 00 00 00 00 10 0f 0e 0d 0c 0b 0a 09 08 07
00e0   06 05 04 03 02 01 00 ac 00 0a 00 00 00 00 00 00
00f0   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

496	24.513295	192.168.1.243	192.168.1.113	TCP	169	2060 → 8000 [PSH, ACK] Seq=511 Ack=1 Win=2048 Len=115
0000   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0010   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0020   00 00 00 00 00 00 00 00 00 00 00 00 06 00 80 04
0030   00 80 17 00 00 0a 00 80 0e 00 00 06 00 00 0c 00
0040   80 00 11 00 00 16 00 80 0d 00 80 a7 7d 03 00 00
0050   00 00 00 00 5e 00 00 e5 00 00 00 02 00 02 00 02
0060   00 02 00 02 00 02 00 02 00 02 13 0a 08 0a 21 1f
0070   ff fe 53

529	27.554389	192.168.1.243	192.168.1.113	TCP	60	2060 → 8000 [FIN, ACK] Seq=626 Ack=1 Win=2048 Len=0
```

**Comment by jkeljo** [comment]
Yeah, the fact that they’re using a streaming protocol (TCP) to send what amount to datagrams has been the cause of more than a little consternation during the development of this library. Thanks for all this great investigation! It’s been a really busy week at work, but I should hopefully have some time to look into all of this over the weekend.

**Comment by jkeljo** [comment]
Wanted to give you an update... took a little longer than I thought to find time to work on this, but I'm just about done rewriting the stream handler to be more robust. I'll put up a new version here when that's done and run with it locally for a bit before pushing a Hass PR to upgrade to it.

**Comment by ptr727** [comment]
Thank you, that would be great.
Any chance you could add the voltage sensor per gem device on the hass side at the same time?

**Comment by jkeljo** [comment]
OK, v1.0.1 of this package has the new parser. You can force your Hass to upgrade to it using the custom deps deployment plugin from the Home Assistant Developer add-ons repository. Add something to the `pypi` section of the config, like so:
```
{
  "pypi": [
    "greeneye-monitor==1.0.1"
  ],
  "apk": []
}
```

If you wouldn't mind running both the Hass component and this library's logging in `debug` mode, like so, that'll help with debugging if you see more problems (setting the `default` to `debug` as you've done has the same effect, but also turns on a bunch of logging from Hass itself, so if you want to avoid that, this will still give all the greeneye-related logs).
```
logger:
  logs:
    homeassistant.components.greeneye: debug
    greeneye.monitor: debug
    greeneye.streams: debug
```

I'll absolutely put up a voltage sensor PR.

**Comment by jkeljo** [comment]
I'm also running it at my house, but you've described your system doing some things that I've never seen mine do (e.g. sending short packets on startup... mine doesn't do that, though it does seem to send malformed packets in the middle of the night regularly), so getting some runtime in a different environment will help.

**Comment by ptr727** [comment]
I run the hass docker container in unraid, and I don't run as root, so I'm struggling a bit with pip and permissions, but I'll figure it out, or I'll just build my own container.

For now I'm running the test app.
I added some logging to log the serial number for all prints, and I added a voltage print.
I notice weird behavior with not all channels printing, sometime all 32 channels will print, sometimes only a few will print.
I've not seen conditional logic to remove unused channels, but I've not looked closely yet.

Modified print logic:
```
def on_new_monitor(monitor):
    monitor.add_listener(lambda: print_monitor(monitor))

    for channel in monitor.channels:
        on_new_channel(monitor, channel)

    for counter in monitor.pulse_counters:
        on_new_counter(monitor, counter)

    for temp in monitor.temperature_sensors:
        on_new_temperature_sensor(monitor, temp)


def on_new_channel(monitor, channel):
    channel.add_listener(lambda: print_channel(monitor, channel))


def on_new_counter(monitor, counter):
    counter.add_listener(lambda: print_counter(monitor, counter))


def on_new_temperature_sensor(monitor, temp):
    temp.add_listener(lambda: print_temperature(monitor, temp))


def print_channel(monitor, channel):
    print("{0} : Channel {1}: {2} W (abs={3} kWh, pol={4} kWh)".format(
        monitor.serial_number,
        channel.number,
        channel.watts,
        channel.absolute_watt_seconds / 3600000,
        channel.polarized_watt_seconds / 3600000))


def print_counter(monitor, counter):
    print("{0} : Pulse counter {1}: {2} ({3}/sec)".format(
        monitor.serial_number,
        counter.number,
        counter.pulses,
        counter.pulses_per_second
    ))


def print_temperature(monitor, sensor):
    print("{0} : Temperature sensor {1}: {2} F".format(
        monitor.serial_number,
        sensor.number,
        sensor.temperature))


def print_monitor(monitor):
    print("{0} : Voltage {1} V".format(
        monitor.serial_number,
        monitor.voltage))
```

Launch.json:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Dump Packets",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/dump_packets.py",
            "console": "integratedTerminal",
            "args": ["8000"]
        }
    ]
}
```

Terminal output:
```
PS C:\Users\piete\source\repos\greeneye-monitor> cd 'c:\Users\piete\source\repos\greeneye-monitor'; ${env:PYTHONIOENCODING}='UTF-8'; ${env:PYTHONUNBUFFERED}='1'; & 'C:\Users\piete\AppData\Local\Programs\Python\Python37\python.exe' 'c:\Users\piete\.vscode-insiders\extensions\ms-python.python-2019.11.45636-dev\pythonFiles\ptvsd_launcher.py' '--default' '--client' '--host' 'localhost' '--port' '53596' 'C:\Users\piete\source\repos\greeneye-monitor/dump_packets.py' '8000' 
2019-11-01 10:04:48,057 [asyncio](DEBUG) Using selector: SelectSelector
2019-11-01 10:04:48,075 [greeneye.monitor](INFO) Server started on ('0.0.0.0', 8000)
2019-11-01 10:04:48,275 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13762)
2019-11-01 10:04:48,281 [greeneye.streams](DEBUG) Received 625 bytes from ('192.168.1.243', 13762)
2019-11-01 10:04:48,289 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:04:48,290 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13762)
2019-11-01 10:04:48,292 [greeneye.monitor](INFO) Discovered new monitor: 1000172
2019-11-01 10:04:48,378 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13778)
2019-11-01 10:04:48,384 [greeneye.streams](DEBUG) Received 625 bytes from ('192.168.1.148', 13778)
2019-11-01 10:04:48,392 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:04:48,406 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13778)
2019-11-01 10:04:48,408 [greeneye.monitor](INFO) Discovered new monitor: 1000181
2019-11-01 10:04:51,512 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13778)
2019-11-01 10:04:51,539 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13762)
2019-11-01 10:04:56,859 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13763)
2019-11-01 10:04:56,889 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13763)
2019-11-01 10:04:56,895 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:04:56,913 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13763)
2019-11-01 10:04:56,919 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:04:56,928 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13763)
2019-11-01 10:04:56,935 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:04:56,936 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13763)
1000172 : Channel 0: 0.0 W (abs=5.555555555555555e-07 kWh, pol=0.0 kWh)
1000172 : Channel 1: 0.0 W (abs=3.611111111111111e-06 kWh, pol=1.388888888888889e-06 kWh)
1000172 : Channel 2: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000172 : Channel 3: 0.0 W (abs=5.555555555555555e-07 kWh, pol=0.0 kWh)
1000172 : Channel 4: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 5: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 6: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 7: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 8: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000172 : Channel 9: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000172 : Channel 10: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000172 : Channel 11: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000172 : Channel 12: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 13: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 14: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 15: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 16: 0.0 W (abs=0.00010972222222222222 kWh, pol=1.6944444444444446e-05 kWh)
1000172 : Channel 17: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 18: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 19: 0.0 W (abs=0.00018194444444444445 kWh, pol=0.00018194444444444445 kWh)
1000172 : Channel 20: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 21: 0.0 W (abs=0.00017166666666666667 kWh, pol=0.00017166666666666667 kWh)
1000172 : Channel 22: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 23: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 24: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 25: 0.0 W (abs=0.00011944444444444444 kWh, pol=0.00011944444444444444 kWh)
1000172 : Channel 26: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 27: 0.0 W (abs=0.00013361111111111112 kWh, pol=0.00013361111111111112 kWh)
1000172 : Channel 28: 0.0 W (abs=8.527777777777777e-05 kWh, pol=8.527777777777777e-05 kWh)
1000172 : Channel 29: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 30: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 31: 0.0 W (abs=0.00016305555555555556 kWh, pol=0.00016305555555555556 kWh)
1000172 : Channel 32: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 33: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 34: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 35: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 36: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 37: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 38: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 39: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 40: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 41: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 42: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 44: 0.0 W (abs=0.0 kWh, pol=19088.743537777777 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000172 : Pulse counter 0: 0 (0.0/sec)
1000172 : Pulse counter 1: 0 (0.0/sec)
1000172 : Pulse counter 2: 1063 (0.0/sec)
1000172 : Pulse counter 3: 4946 (0.0/sec)
1000172 : Voltage 120.2 V
2019-11-01 10:04:56,984 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13779)
2019-11-01 10:04:57,015 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13779)
2019-11-01 10:04:57,016 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:04:57,038 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13779)
2019-11-01 10:04:57,039 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:04:57,052 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13779)
2019-11-01 10:04:57,053 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:04:57,054 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13779)
1000181 : Channel 0: 0.1 W (abs=2.0865797222222224 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 224.3 W (abs=1220.0675261111112 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 2: 0.0 W (abs=2.5e-06 kWh, pol=2.7777777777777776e-07 kWh)
1000181 : Channel 3: 0.0 W (abs=2.777777777777778e-06 kWh, pol=2.7777777777777776e-07 kWh)
1000181 : Channel 4: 0.0 W (abs=8.333333333333333e-07 kWh, pol=0.0 kWh)
1000181 : Channel 5: 0.0 W (abs=8.333333333333333e-07 kWh, pol=0.0 kWh)
1000181 : Channel 6: 0.0 W (abs=8.333333333333333e-07 kWh, pol=0.0 kWh)
1000181 : Channel 7: 0.0 W (abs=8.333333333333333e-07 kWh, pol=0.0 kWh)
1000181 : Channel 8: 0.0 W (abs=1.388888888888889e-06 kWh, pol=8.333333333333333e-07 kWh)
1000181 : Channel 9: 0.0 W (abs=2.222222222222222e-06 kWh, pol=1.388888888888889e-06 kWh)
1000181 : Channel 10: 0.0 W (abs=3.611111111111111e-06 kWh, pol=3.0555555555555556e-06 kWh)
1000181 : Channel 11: 0.0 W (abs=1.9444444444444444e-06 kWh, pol=1.388888888888889e-06 kWh)
1000181 : Channel 12: 0.0 W (abs=1.388888888888889e-06 kWh, pol=1.388888888888889e-06 kWh)
1000181 : Channel 13: 0.0 W (abs=1.388888888888889e-06 kWh, pol=1.388888888888889e-06 kWh)
1000181 : Channel 14: 0.0 W (abs=0.0013844444444444444 kWh, pol=8.333333333333333e-07 kWh)
1000181 : Channel 15: 0.0 W (abs=0.0022608333333333334 kWh, pol=2.7777777777777776e-07 kWh)
1000181 : Channel 16: 0.0 W (abs=0.0004375 kWh, pol=0.00018583333333333334 kWh)
1000181 : Channel 17: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000181 : Channel 18: 0.0 W (abs=5.555555555555555e-07 kWh, pol=0.0 kWh)
1000181 : Channel 19: 0.0 W (abs=2.7777777777777776e-07 kWh, pol=0.0 kWh)
1000181 : Channel 20: 0.0 W (abs=0.0018863888888888888 kWh, pol=5.555555555555555e-07 kWh)
1000181 : Channel 21: 0.0 W (abs=0.00028 kWh, pol=2.7777777777777776e-07 kWh)
1000181 : Channel 22: 0.0 W (abs=0.0020516666666666665 kWh, pol=0.0 kWh)
1000181 : Channel 23: 0.0 W (abs=0.002043333333333333 kWh, pol=0.0 kWh)
1000181 : Channel 24: 0.0 W (abs=8.583333333333334e-05 kWh, pol=8.5e-05 kWh)
1000181 : Channel 25: 0.0 W (abs=0.001926388888888889 kWh, pol=8.472222222222222e-05 kWh)
1000181 : Channel 26: 0.0 W (abs=8.5e-05 kWh, pol=8.5e-05 kWh)
1000181 : Channel 27: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 28: 0.0 W (abs=1.6666666666666667e-06 kWh, pol=1.388888888888889e-06 kWh)
1000181 : Channel 29: 0.0 W (abs=0.0008377777777777778 kWh, pol=5.555555555555555e-07 kWh)
1000181 : Channel 30: 0.0 W (abs=8.333333333333333e-07 kWh, pol=5.555555555555555e-07 kWh)
1000181 : Channel 31: 0.0 W (abs=3.0555555555555556e-06 kWh, pol=1.111111111111111e-06 kWh)
1000181 : Channel 32: -5767161.6 W (abs=0.0 kWh, pol=3.9503644444444443 kWh)
1000181 : Channel 33: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 34: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 35: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 36: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 37: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 38: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 39: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 40: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 41: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 42: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 44: 0.0 W (abs=0.0 kWh, pol=19088.743537777777 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000181 : Pulse counter 0: 0 (0.0/sec)
1000181 : Pulse counter 1: 0 (0.0/sec)
1000181 : Pulse counter 2: 0 (0.0/sec)
1000181 : Pulse counter 3: 0 (0.0/sec)
1000181 : Voltage 122.1 V
2019-11-01 10:05:00,547 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13779)
2019-11-01 10:05:00,575 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13763)
2019-11-01 10:05:06,860 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13764)
2019-11-01 10:05:06,888 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13764)
2019-11-01 10:05:06,890 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:06,913 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13764)
2019-11-01 10:05:06,913 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:06,927 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13764)
2019-11-01 10:05:06,929 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:06,929 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13764)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Voltage 120.3 V
2019-11-01 10:05:06,984 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13780)
2019-11-01 10:05:07,015 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13780)
2019-11-01 10:05:07,025 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:07,039 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13780)
2019-11-01 10:05:07,055 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:07,056 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13780)
2019-11-01 10:05:07,058 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:07,059 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13780)
1000181 : Channel 0: 0.5 W (abs=2.086581111111111 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 229.0 W (abs=1220.0681622222223 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 32: -104857.6 W (abs=0.0 kWh, pol=4.096 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Voltage 122.1 V
2019-11-01 10:05:10,588 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13780)
2019-11-01 10:05:10,616 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13764)
2019-11-01 10:05:16,859 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13765)
2019-11-01 10:05:16,890 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13765)
2019-11-01 10:05:16,892 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:16,912 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13765)
2019-11-01 10:05:16,920 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:16,928 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13765)
2019-11-01 10:05:16,934 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:16,935 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13765)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 44: 0.0 W (abs=0.0 kWh, pol=19088.743537777777 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000172 : Voltage 120.2 V
2019-11-01 10:05:16,986 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13781)
2019-11-01 10:05:17,016 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13781)
2019-11-01 10:05:17,017 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:17,040 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13781)
2019-11-01 10:05:17,065 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:17,069 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13781)
2019-11-01 10:05:17,074 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:17,087 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13781)
1000181 : Channel 0: 0.0 W (abs=2.086581111111111 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 228.2 W (abs=1220.0687961111112 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 32: -39321.6 W (abs=0.0 kWh, pol=4.150613333333333 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000181 : Voltage 122.0 V
2019-11-01 10:05:20,627 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13781)
2019-11-01 10:05:20,656 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13765)
2019-11-01 10:05:26,860 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13766)
2019-11-01 10:05:26,889 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13766)
2019-11-01 10:05:26,893 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:26,913 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13766)
2019-11-01 10:05:26,916 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:26,928 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13766)
2019-11-01 10:05:26,931 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:26,932 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13766)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000172 : Voltage 119.9 V
2019-11-01 10:05:26,979 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13782)
2019-11-01 10:05:27,011 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13782)
2019-11-01 10:05:27,012 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:27,036 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13782)
2019-11-01 10:05:27,043 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:27,049 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13782)
2019-11-01 10:05:27,062 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:27,063 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13782)
1000181 : Channel 0: 0.4 W (abs=2.0865822222222223 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 226.8 W (abs=1220.0694261111112 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 32: -26214.6 W (abs=0.0 kWh, pol=4.1870225 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
1000181 : Voltage 122.1 V
2019-11-01 10:05:30,668 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13782)
2019-11-01 10:05:30,696 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13766)
2019-11-01 10:05:36,859 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13767)
2019-11-01 10:05:36,889 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13767)
2019-11-01 10:05:36,893 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:36,913 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13767)
2019-11-01 10:05:36,919 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:36,928 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13767)
2019-11-01 10:05:36,934 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:36,934 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13767)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Voltage 119.9 V
2019-11-01 10:05:36,962 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13783)
2019-11-01 10:05:36,992 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13783)
2019-11-01 10:05:36,993 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:37,016 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13783)
2019-11-01 10:05:37,020 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:37,030 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13783)
2019-11-01 10:05:37,039 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:37,040 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13783)
1000181 : Channel 0: 0.3 W (abs=2.0865830555555553 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 226.5 W (abs=1220.0700552777778 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 32: -5963770.0 W (abs=0.0 kWh, pol=4.096000277777778 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Voltage 122.1 V
2019-11-01 10:05:40,708 [greeneye.streams](INFO) Connection closed from ('192.168.1.148', 13783)
2019-11-01 10:05:40,737 [greeneye.streams](INFO) Connection closed from ('192.168.1.243', 13767)
2019-11-01 10:05:46,859 [greeneye.streams](INFO) Connection from ('192.168.1.243', 13768)
2019-11-01 10:05:46,889 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13768)
2019-11-01 10:05:46,891 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:46,914 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.243', 13768)
2019-11-01 10:05:46,914 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:46,927 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.243', 13768)
2019-11-01 10:05:46,929 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:46,929 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.243', 13768)
1000172 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000172 : Channel 44: 0.0 W (abs=0.0 kWh, pol=19088.743537777777 kWh)
1000172 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000172 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000172 : Channel 47: 0.0 W (abs=0.0 kWh, pol=1202.4220458333334 kWh)
2019-11-01 10:05:46,941 [greeneye.streams](INFO) Connection from ('192.168.1.148', 13784)
2019-11-01 10:05:46,971 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13784)
2019-11-01 10:05:46,972 [greeneye.streams](DEBUG) Not enough data in buffer yet (255 bytes)
2019-11-01 10:05:46,996 [greeneye.streams](DEBUG) Received 255 bytes from ('192.168.1.148', 13784)
2019-11-01 10:05:46,997 [greeneye.streams](DEBUG) Not enough data in buffer yet (510 bytes)
2019-11-01 10:05:47,010 [greeneye.streams](DEBUG) Received 115 bytes from ('192.168.1.148', 13784)
2019-11-01 10:05:47,013 [greeneye.streams](DEBUG) Parsed one BIN48-NET-TIME packet.
2019-11-01 10:05:47,014 [greeneye.streams](DEBUG) Parsed 1 packet from ('192.168.1.148', 13784)
1000181 : Channel 0: 0.1 W (abs=2.0865833333333335 kWh, pol=1.9109966666666667 kWh)
1000181 : Channel 1: 222.1 W (abs=1220.0706722222221 kWh, pol=0.7640688888888889 kWh)
1000181 : Channel 32: -5885126.6 W (abs=0.0 kWh, pol=4.077795555555555 kWh)
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
1000181 : Voltage 122.0 V
2019-11-01 10:05:48,078 [greeneye.monitor](INFO) Closing server on ('0.0.0.0', 8000)
PS C:\Users\piete\source\repos\greeneye-monitor>
```

I can point my sensors at you, and you can log your own internal and my two sensors if you want to test with multiple GEM's?

**Comment by jkeljo** [comment]
That's looking pretty good.

The conditional logic is actually in the Channel, TemperatureSensor, and PulseCounter objects; they only notify listeners when the packet causes their internal state to change. If a channel reports no difference in wattseconds between two packets, it won't notify.

No need to point your stuff at me; I realized while reacquainting myself with the code that it should already handle multiple GEMs. It's been running for 24 hours here with no issue, so I'll put up the upgrade PR this weekend.

**Comment by ptr727** [comment]
Ah, I found the comparison logic in the objects.

What is the reason for not reporting out when the same value is reported by the GEM?

I wonder how the sampling will look when I log data over time, how do I know a value is missing from the sample set, or a value was the same as the previous sample?
E.g. reporting the same temperature over and over is expected is the temperature was the same at every interval.

I do see how calculating rate changes for pule counters is valuable, but reporting the same value at every sample is again useful for continuous monitoring.

For the current sensing, the GEM calculates the absolute consumption itself, and like pulse counters you calculate useful rate changes, but as before, why not report the value as sampled, even if rate changes are then 0?

In my setup serial 1000181 should be reporting on channel 0 and 1, and serial 1000172 should be reporting on pulse counter 0 and 1. Any other reports on channels are noise. In the GEM utility one can see the values on empty channels reported as 0, -0, 0, 0, -0, etc.

The GEM's only have 32 channels, yet e.g. channels 43, 45, and 46 report changes, but the values appear to not change, or the conditional logic uses non printing value comparisons?


```
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
```

```
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
```

```
1000181 : Channel 43: 0.0 W (abs=0.0 kWh, pol=0.0 kWh)
1000181 : Channel 45: 0.0 W (abs=0.0 kWh, pol=13179.672893055556 kWh)
1000181 : Channel 46: 0.0 W (abs=0.0 kWh, pol=7191.047469444445 kWh)
```

**Comment by jkeljo** [comment]
Yeah, the conditional logic for firing a change listener on the channels also looks at the reported current (which the test app does not print), so that's probably what's going on.

As to why not fire the listeners every time a packet came in... at the level of the `CurrentSensor`, `TemperatureSensor`, etc. in this API, these are change listeners. For a client that cares only about displaying the latest values, this is ideal because it minimizes the amount of work required to keep the display up to date.

If a client wants to know about every packet, it can listen to the change listener on the `Monitor`, which fires on every packet. However, I don't think it's necessary. Any client that's recording this data over time will need to be able to deal with uneven gaps between data points, because of a variety of causes, including local network conditions, differences in the load on the machine receiving the data, and whatever is causing the GEM to send the occasional malformed packet. In the case of the Home Assistant integration, InfluxDB and Grafana both have query and rendering modes that extrapolate from samples by assuming the previous value holds until the next sample, for example.

**Comment by ptr727** [comment]
I looked at the binary details of a few of the weird channels, and the data the GEM sends is just weird, so not an issue.

I do however feel that not reporting on valid data makes it impossible to know if the sensor was online or offline. Yes, one can extrapolate data, but absent data could mean a missed / corrupt report interval, an offline system / power failure, or the same value, how would I know?

I would much rather the system report data as it is received, vs. assuming the data is not useful or could be recovered.

Would you please consider adding a configuration option, also exposed in HASS, that allows for not suppressing (default behavior as today) same value reports and reporting whenever data is received?

I notice the temperature values seem to be doubled, per the docs I think they need to be divided by 2. I don't have temp sensors attached so I can't really tell, and the app detects missing sensors while the value reads as 512.

> TEMPERATURE
> The temperature value is taken from its assigned sensor
> If using an HTTP format, the temperature value outputted is the raw value from the sensor.
> If you’re using Binary, the value is multiplied by 2 due to the 0.5 C resolution of the DS18B20. When the value is parsed via software, it must be divided by 2 to obtain its value.
>

**Comment by jkeljo** [comment]
> Would you please consider adding a configuration option, also exposed in HASS, that allows for not suppressing (default behavior as today) same value reports and reporting whenever data is received?

There's already enough support for direct callers of this API to do that themselves (subscribe to the listener on the `Monitor`). Unfortunately it's not currently possible to plumb that all the way through Home Assistant to InfluxDB, since Home Assistant itself only fires state change notifications when a sensor value changes. What I can do, though, is add a sensor that represents the status of the monitor itself somehow.

Good catch on the temperature value. I don't have temperature sensors hooked up, either, but the docs are pretty clear so I'll just fix that.

> the app detects missing sensors

Which app?

**Comment by ptr727** [comment]
App, the GEM Network Utility and the web live data page shows the temp sensors as missing / not present. I think I have some 1-wire probes somewhere, I'll get one and hook it up and see how the temp readings change from 512/2.0 to something else.

For periodic vs. state change reporting, if your code works similar to how hass handles other sensor type data, then ok.
But, as far as I can see from graph data, my weather station and temp sensor data is recorded every time the sensor reports, not just when the value changes from the previous value. Now I may be wrong here, and maybe it just looks like the graph shows continuous recording?

The reason why understanding the use of continuous vs. value change logging is important to me, is because I do that for all my other sensor sources (weather, solar, power quality, water, gas, etc.), and I was planning of consolidating all to hass, instead of my own various pull to log, insert into DB tools I use.

**Comment by jkeljo** [comment]
https://github.com/home-assistant/home-assistant/pull/28600 upgrades Home Assistant to the new version of this library with the crash fix and revamped packet handler.

**Comment by ptr727** [comment]
Thank you.
I got some temp probes, I'll install the weekend and let you know how the reported values change.
Does not look like the voltage sensor has been added yet?

**Comment by jkeljo** [comment]
No, their turnaround on PRs is usually a week or so to merge, then another couple weeks waiting for the next release. You can always apply the PR locally and put the component in the custom_components directory to get Hass to pick it up. (I run that way a lot while waiting for PRs to make their way thru the process.)

**Comment by ptr727** [comment]
I'll try out custom_components, running docker hass instead of vm hass.io makes life a bit complicated sometimes.

From the diff the PR updates the dependency version, I did not see a change to add a voltage sensor in hass, or did I miss that?

**Comment by jkeljo** [comment]
There’s two PRs, one for the upgrade that I put in yesterday, and one for the voltage sensor that I put up last week.

**Comment by ptr727** [comment]
As FYI, I installed a temp sensor, set to degC in config.
The reading on the installed sensor is 19.5, the reading on the missing sensors are 256, this is after dividing the raw input by 2.0.

**Comment by ptr727** [comment]
Seems they have still not merged your changes?

**Comment by ptr727** [comment]
I don't want to ruffle feathers, but this is getting a bit long in waiting for a merge, what can I do to help things move along?

**Comment by jkeljo** [comment]
Yeah, it's frustrating. Looks like another PR came in to do more or less the same thing and was somehow quickly accepted. I'm not sure what the difference was that made that one go fast.

My attempted fix for this particular issue was in Hass v0.102. Did that address the problem for you?

**Comment by ptr727** [comment]
That change got in and accepted in no time, looks no different to yours that's been waiting 2 months :(
I'll test with 102, thank you.

**Comment by zaped212** [comment]
Is this expected to be resolved in HA 0.104 ?
I have been having what seems to be this issue as well, though I only have one GEM, but I do have 30 sensors.

**Comment by jkeljo** [comment]
Yes it is. Please file a new issue here with your logs; could be something
else is happening.

**Comment by zaped212** [comment]
@jkeljo I am trying to nail it down exactly first. Though things seem more stable with the greeneye_monitor code disabled. I will let my system run a day with it off and everything else on to confirm stability, and then will see about collecting logs and opening up a new issue.

**Comment by jkeljo** [comment]
I'm going to close this issue now. If either of you encounter problems on HA 0.102 or later, please file a new issue with your logs.

**Comment by zaped212** [comment]
Thanks @jkeljo it looks like I am not having any issues caused by this plugin.
